### PR TITLE
06e <- 07a Flask Updates (Do Not Merge)

### PR DESCRIPTION
**Pull request by kelsey-steven-ada**

### CODE REVIEW

File: tests/test_book_routes.py [hunk 1]
```diff
@@ -28,6 +92,43 @@ def test_create_one_book(client):
...
+        "extra": "some stuff",
+        "title": "New Book",
+        "description": "The Best!",
+        "another": "last value"
+    }
+
+    # Act
+    response = client.post("/books", json=test_data)
+    response_body = response.get_json()
+
```
Comment (anselrognlie) [inline tests/test_book_routes.py]: Shouldn't these not be added until 07b? Learn introduces them along with the from_dict refactor: https://github.com/Ada-Developers-Academy/core-unit-2/blob/main/api-7-refactoring/extracting-helper-functions-from-dict.md

File: tests/test_book_routes.py [hunk 2]
```diff
@@ -28,6 +92,43 @@ def test_create_one_book(client):
...
+        "description": "The Best!"
+    }
+
+def test_create_one_book_no_title(client):
+    # Arrange
+    test_data = {"description": "The Best!"}
+
+    # Act & Assert
+    with pytest.raises(KeyError, match="title"):
+        response = client.post("/books", json=test_data)
```
Comment (anselrognlie) [inline tests/test_book_routes.py]: Aside from potentially moving these tests to 07b, this really surprised me. I would have expected the response to just give backthe typical 500 internal server error (like we see through postman) rather than exposing the KeyError. Is this behavior just occuring due to using the flask test client, and the abstraction is a little leaky?

**Review by anselrognlie — CHANGES_REQUESTED** [review]
Looks good overall, but shouldn't the additional create tests not be added yet?

File: tests/test_book_routes.py (hunk 1 above)
Comment (kelsey-steven-ada) [reply tests/test_book_routes.py]: This came about because the create route now returns a Book record instead of a string. The create route is a dependency for the `to_dict` refactor, which is why I moved these tests up.

Comment (anselrognlie) [reply tests/test_book_routes.py]: Ah, I see. Would it be strange to only add the successful test cases here (the ones that actually make it through to returning the dict)? Per my other comment, having the `KeyError` tests is a little odd here. And if we're going to immediately turn around and fix them for the message output, it seems unnecessary to get into those weeds here.

File: tests/test_book_routes.py (hunk 2 above)
Comment (kelsey-steven-ada) [reply tests/test_book_routes.py]: We do get a 500 internal server error when hitting the endpoint in Postman. If we are in production mode Flask hides the KeyError, but in debug mode, it sends the KeyError and stack trace to Postman. 

I believe this is the test client, the test exits as soon as the code crashes and the KeyError gets raised. If we instead tried to write something like:
```python
test_data = {"title": "New Book"}
response = client.post("/books", json=test_data)
assert response.status_code == 500  
```
The test fails as soon as we hit the KeyError and we never get to the assert.

Comment (anselrognlie) [reply tests/test_book_routes.py]: That's why this feels leaky to me. Like it feels very strange to write a test that's verifying debug behavior. Essentially, I'm advocating for a way we can avoid showing this, such as moving the failure cases to the next branch where we have better error handling.

Comment (kelsey-steven-ada) [reply tests/test_book_routes.py]: This may feel "leaky" but it is how Flask is meant to work. I dug into the app and testing config, and this is caused by us setting `"TESTING": True`: https://flask.palletsprojects.com/en/3.0.x/config/#TESTING. If we don't set the testing flag, then we could write a test that checks for a generic 500 error response.

To me, this makes sense for testing; if I'm working on my app, getting a generic 500 error from this test doesn't help me debug an issue nearly as much as knowing there's an error being raised.

File: tests/test_book_routes.py (hunk 1 above)
Comment (kelsey-steven-ada) [reply tests/test_book_routes.py]: I don't feel that it's a great example to purposefully leave off edge cases when we're making changes that could be affected by those edge cases. My thoughts are that a project's test suite has to evolve with the code over time, and it isn't a bad thing to show off that sometimes we do need to revisit tests because we made a change to the logic to make our app stronger overall. 

In order to keep our lessons focused, I don't really want to add changes to code that isn't otherwise handled in this lesson. The next lesson handles Book creation, so it makes sense to me to leave error handling for the Book creation bundled with that lesson. I'd want to get other folk's thoughts, but if you think it's a detriment to students, what do you think about swapping the order of the lessons so the `from_dict` refactor comes first?

File: tests/test_book_routes.py (hunk 2 above)
Comment (anselrognlie) [reply tests/test_book_routes.py]: That's true. It's useful to see the underlying error if it occurs while running a test (now that I think of it, I do see this behavior while working), but I think that's more because it shows us there's a place where an exception _unexpectedly_ escaped. In that case, I would go in and write code to handle the exception (or otherwise figure out why the exception happened) rather than adding a test to receive that exception.

File: tests/test_book_routes.py (hunk 1 above)
Comment (anselrognlie) [reply tests/test_book_routes.py]: Oh, I like that. We can implement the `from_dict`, but still leave the existing repetative return value, then fix the `to_dict` in the next section. That would allow us to skip the `KeyError` issue pretty neatly.

File: tests/test_book_routes.py (hunk 2 above)
Comment (kelsey-steven-ada) [reply tests/test_book_routes.py]: That is fair that we'd want to go in and fix that rather than just testing for it. While we could talk around that in class and say we'll address it soon, swapping the lesson order very neatly removes the need for that and any confusion it could cause.

Comment (anselrognlie) [reply tests/test_book_routes.py]: I agree. I think you identified the better solution.

**Review by anselrognlie — APPROVED** [review]
LGTM
